### PR TITLE
[crypto] Add Pedersen Commitments and Range Proofs

### PR DESCRIPTION
**Pull request by punwai**
This PR exposes new APIs for Pedersen Commitments and Bulletproofs range proofs. We expose the functionality to:

- Create Pedersen Commitment for some value and blinding factor
- Add and subtract Pedersen Commitments
- Create a Range Proof for some value and blinding factor
- Verify a Range Proof for some Pedersen commitment, given the range

The APIs will be exposed as native functions in Sui Move.

### CODE REVIEW

File: crypto/src/tests/secp256k1_tests.rs [hunk 1]
```diff
@@ -258,11 +258,11 @@ proptest::proptest! {
 
         // same recovered pubkey are recovered
         let recovered_key = signature.sig.recover(&hashed_msg).unwrap();
-        let recovered_key_1 = signature_1.recover_verifying_key(message).expect("couldn't recover pubkey");
+        let recovered_key_1 = signature_1.recover_verify_key(message).expect("couldn't recover pubkey");
         assert_eq!(recovered_key.serialize(),recovered_key_1.to_bytes().as_slice());
 
-        // same signatures produced from both implementations
-        assert_eq!(signature.as_ref(), ToFromBytes::as_bytes(&signature_1));
+        // // same signatures produced from both implementations
```
Comment (punwai) [inline crypto/src/tests/secp256k1_tests.rs:264]: Changes in versioning changes k256's signing implementation.

File: crypto/src/bulletproofs.rs [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+        let pc_gens = PedersenGens::default();
+        let bp_gens = BulletproofGens::new(64, 1);
+        let mut prover_transcript = Transcript::new(b"doctest example");
+
+        let blinding = Scalar::from_bits(blinding);
+        let bounding: u64 = u64::max_value();
+        let new_value = value + (bounding - upper_bound);
+
+        // Create a 64-bit rangeproof.
+        let (proof, _) = RangeProof::prove_single(
```
Comment (punwai) [inline crypto/src/bulletproofs.rs:154]: Every proof is created in 0 ... 64 range

File: crypto/src/bulletproofs.rs [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+
+impl AsRef<[u8]> for PedersenCommitment {
+    fn as_ref(&self) -> &[u8] {
+        self.bytes.get_or_init(|| self.point.compress().to_bytes())
+    }
+}
+
+impl ToFromBytes for PedersenCommitment {
+    fn from_bytes(bytes: &[u8]) -> Result<Self, signature::Error> {
+        if bytes.len() != 32 {
```
Comment (akichidis) [inline crypto/src/bulletproofs.rs:71]: Could you please extract the number `32` to a const so we avoid [magic numbers](https://www.codebyamir.com/blog/software-anti-patterns-magic-numbers#:~:text=A%20magic%20number%20is%20a,so%20they%20should%20be%20avoided.) ?

File: crypto/src/bulletproofs.rs [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+}
+
+impl BulletproofsRangeProof {
+    // Prove that commited value is within the range [0, 2^64)
+    pub fn prove_bound(
+        value: u64,
+        blinding: [u8; 32],
+    ) -> Result<(PedersenCommitment, Self), signature::Error> {
+        let pc_gens = PedersenGens::default();
+        let bp_gens = BulletproofGens::new(64, 1);
```
Comment (akichidis) [inline crypto/src/bulletproofs.rs:141]: Same comment as above, I would suggest extracting value `64` to a const

File: crypto/src/bulletproofs.rs [hunk 5]
```diff
@@ -0,0 +1,263 @@
...
+        let mut prover_transcript = Transcript::new(b"mysten");
```
Comment (akichidis) [inline crypto/src/bulletproofs.rs:142]: Not familiar with the `Transcript` but as far as I understood this is being used for domain separation. In this case do we want to go with `mysten` or rather the specific project (ie `narwhal`). I mean is this effectively allow the proofs to work across our products (ex assuming SUI is also using the same domain). We might (or not) want that. Apologies if my observation is totally off - not familiar with the  foundation.

File: crypto/src/tests/secp256k1_tests.rs [hunk 6]
```diff
@@ -258,11 +258,11 @@ proptest::proptest! {
 
         // same recovered pubkey are recovered
         let recovered_key = signature.sig.recover(&hashed_msg).unwrap();
-        let recovered_key_1 = signature_1.recover_verifying_key(message).expect("couldn't recover pubkey");
+        let recovered_key_1 = signature_1.recover_verify_key(message).expect("couldn't recover pubkey");
```
Comment (akichidis) [inline crypto/src/tests/secp256k1_tests.rs:261]: it seems that some tests fail because of the renaming

**Review by akichidis — APPROVED** [review]
Additions LGTM - left a few comments. I am approving but I would strongly suggest to wait for someone more experienced in cryptography to also approve before merging.

File: crypto/src/bulletproofs.rs (hunk 5 above)
Comment (punwai) [reply crypto/src/bulletproofs.rs:142]: Yes, I believe this is domain separation - I'm happy to go with narwhal for this PR, but I think this entire crate is actually going to be moved to mysten-infra

File: crypto/src/tests/secp256k1_tests.rs (hunk 6 above)
Comment (punwai) [reply crypto/src/tests/secp256k1_tests.rs:261]: Yep, this is reverted!

File: crypto/src/bulletproofs.rs (hunk 3 above)
Comment (kchalkias) [reply crypto/src/bulletproofs.rs:71]: Yes let's encode this as const ped commitment length

File: crypto/src/bulletproofs.rs [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+#[derive(Debug)]
+pub struct BulletproofsRangeProof {
+    proof: RangeProof,
+    bytes: OnceCell<Vec<u8>>,
+}
+
+impl BulletproofsRangeProof {
+    /// Prove that commited value is within the range [0, 2^64). Function only works
+    /// for n = 8, 16, 32, 64
+    pub fn prove_bound(
```
Comment (kchalkias) [inline crypto/src/bulletproofs.rs:137]: can you elaborate what `n` is here? if it's the power of two then it's conflicting with the message that says within [0, 2^64) 
-> it should be [0, 2^n)

File: crypto/src/bulletproofs.rs [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+        value: u64,
+        blinding: [u8; 32],
+        n: usize,
+        domain: &'static [u8]
+    ) -> Result<(PedersenCommitment, Self), signature::Error> {
+        // Although this is also checked in the bulletproofs library, we check again
+        // to avoid unexpected behaviour in the case of library updates
+        if !(n == 8 || n == 16 || n == 32 || n == 64) {
+            return Err(signature::Error::new());
```
Comment (kchalkias) [inline crypto/src/bulletproofs.rs:146]: Better error msg? No worries just add a TODO and we can add in an upcoming PR for all errors in this commit

File: crypto/src/bulletproofs.rs [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+                bytes: OnceCell::new(),
+            },
+            BulletproofsRangeProof {
+                proof,
+                bytes: OnceCell::new(),
+            },
+        ))
+    }
+
+    // Verify that commited value is within the range [0, 2^64)
```
Comment (kchalkias) [inline crypto/src/bulletproofs.rs:176]: same: 2^n

File: crypto/src/bulletproofs.rs [hunk 10]
```diff
@@ -0,0 +1,218 @@
...
+    ) -> Result<(), signature::Error> {
+        if !(n == 8 || n == 16 || n == 32 || n == 64) {
+            return Err(signature::Error::new());
+        }
+
+        let pc_gens = PedersenGens::default();
+        let bp_gens = BulletproofGens::new(n, 1);
+        let mut verifier_transcript = Transcript::new(domain);
+
+        // Create a 64-bit rangeproof.
```
Comment (kchalkias) [inline crypto/src/bulletproofs.rs:191]: does it depend on n? or it's always 64?

File: crypto/src/bulletproofs.rs (hunk 7 above)
Comment (punwai) [reply crypto/src/bulletproofs.rs:137]: That's my error - I didn't see this comment

File: crypto/src/bulletproofs.rs (hunk 10 above)
Comment (punwai) [reply crypto/src/bulletproofs.rs:191]: Yes, this depends on n - I need to start reading the comments I write before submitting PRs. I've been ignoring them and looking at the code in the past.

File: crypto/src/bulletproofs.rs (hunk 9 above)
Comment (punwai) [reply crypto/src/bulletproofs.rs]: Fixed all

File: crypto/src/bulletproofs.rs [hunk 11]
```diff
@@ -0,0 +1,229 @@
+use std::ops;
```
Comment (kchalkias) [inline crypto/src/bulletproofs.rs:1]: add license header

**Review by kchalkias — APPROVED** [review]
Awesome, this is a good start towards privacy-related api, will approve!

**Pull request merged**